src/sentry/integrations/messaging.py

```diff
@@ -216,3 +237,174 @@ def build_handler(
+
+    @method_decorator(never_cache)
+    def handle(
+        self, request: HttpRequest, signed_params: Any, *args: Any, **kwargs: Any
+    ) -> HttpResponseBase:
+        try:
+            params = unsign(signed_params, salt=self.salt)
+        except (SignatureExpired, BadSignature):
+            return render_to_response(
+                self.expired_link_template,
+                request=request,
+            )
+
+        organization: RpcOrganization | None = None
+        integration: Integration | None = None
+        idp: IdentityProvider | None = None
+        integration_id = params.get("integration_id")
+        if integration_id:
+            organization, integration, idp = get_identity_or_404(
+                self.provider,
+                request.user,
+                integration_id=integration_id,
+                organization_id=params.get("organization_id"),
+            )
+
+        if request.method != "POST":
+            context = {
+                "organization": organization,
+                "provider": integration.get_provider() if integration else None,
+            }
+            return render_to_response(self.confirmation_template, request=request, context=context)
+
+        response = self.execute(idp, params, request)
+        if response is not None:
+            return response
+
+        self.notify_on_success(integration, params)
+
+        if self.success_metric:
+            provider_slug = self.parent_messaging_spec.provider_slug
+            analytics.record(
+                self.success_metric,
+                provider=provider_slug,
+                actor_id=request.user.id,
+                actor_type=ActorType.USER,
+            )
+
+        team_id = params.get("team_id")
+        context = {"team_id": team_id} if team_id else {}
+        return render_to_response(self.success_template, request=request, context=context)
+
+    @abstractmethod
+    def execute(
+        self, idp: IdentityProvider | None, params: Mapping[str, Any], request: HttpRequest
+    ) -> HttpResponse | None:
+        """Execute the operation on the Identity table.
+
+        Return a response to trigger an early halt under exceptional conditions.
+        Return None if everything is normal.
+        """
+        raise NotImplementedError
+
+
+class LinkIdentityView(LinkingView, ABC):
+    @property
+    def confirmation_template(self) -> str:
+        return "sentry/auth-link-identity.html"
+
+    def execute(
+        self, idp: IdentityProvider | None, params: Mapping[str, Any], request: HttpRequest
+    ) -> HttpResponse | None:
+        if idp is None:
+            raise ValueError('idp is required for linking (params must include "integration_id")')
+
+        user = request.user
+        if isinstance(user, AnonymousUser):
+            raise TypeError("Cannot link identity without a logged-in user")
+
+        Identity.objects.link_identity(user=user, idp=idp, external_id=params[self.user_parameter])
```

Comment:
should this be put in a try/except and should we track metrics for it? i'm not sure how much identity linking errors relate to system or user errors

Reply:
We got this from the Slack code when https://github.com/getsentry/sentry/pull/75899 brought it in. Failures on this query are now captured as metrics for all providers.